src/libs/sfdk/vboxvirtualmachine.cpp

```diff
@@ -0,0 +1,1008 @@
+    sargs.append(QLatin1String(YES_ARG));
+
+    auto enqueue = [=](const QStringList &args, CommandQueue::BatchId batch) {
+        auto runner = std::make_unique<VBoxManageRunner>(args);
+        QObject::connect(runner.get(), &VBoxManageRunner::failure, Sdk::instance(), [=]() {
+            commandQueue()->cancelBatch(batch);
+            callIf(context_, functor, false);
+        });
+        VBoxManageRunner *runner_ = runner.get();
+        commandQueue()->enqueue(std::move(runner));
+        return runner_;
+    };
+
+    CommandQueue::BatchId batch = commandQueue()->beginBatch();
+    enqueue(rargs, batch);
+    enqueue(aargs, batch);
+    enqueue(sargs, batch);
+    commandQueue()->enqueueCheckPoint(context, std::bind(functor, true));
+    commandQueue()->endBatch();
+}
+
+void VBoxVirtualMachinePrivate::doAddPortForwarding(const QString &ruleName,
+        const QString &protocol, quint16 hostPort, quint16 emulatorVmPort,
+        const QObject *context, const Functor<bool> &functor)
+{
+    Q_Q(VBoxVirtualMachine);
+    Q_ASSERT(context);
+    Q_ASSERT(functor);
+
+    doRemovePortForwarding(ruleName, Sdk::instance(), [](bool) {/* noop */});
+
+    qCDebug(vms) << "Setting port forwarding for" << q->uri().toString() << "from"
+        << hostPort << "to" << emulatorVmPort;
+
+    QStringList arguments;
+    arguments.append(QLatin1String(MODIFYVM));
+    arguments.append(q->name());
+    arguments.append(QLatin1String(NATPF1));
+    arguments.append(QString::fromLatin1("%1,%2,,%3,,%4")
+                     .arg(ruleName).arg(protocol).arg(hostPort).arg(emulatorVmPort));
+
+    auto runner = std::make_unique<VBoxManageRunner>(arguments);
+    QObject::connect(runner.get(), &VBoxManageRunner::done, context, functor);
+    commandQueue()->enqueueImmediate(std::move(runner));
+}
+
+void VBoxVirtualMachinePrivate::doRemovePortForwarding(const QString &ruleName,
+        const QObject *context, const Functor<bool> &functor)
+{
+    Q_Q(VBoxVirtualMachine);
+    Q_ASSERT(context);
+    Q_ASSERT(functor);
+
+    qCDebug(vms) << "Deleting port forwarding rule" << ruleName << "from" << q->uri().toString();
+
+    QStringList arguments;
+    arguments.append(QLatin1String(MODIFYVM));
+    arguments.append(q->name());
+    arguments.append(QLatin1String(NATPF1));
+    arguments.append(QLatin1String(DELETE));
+    arguments.append(ruleName);
+
+    auto runner = std::make_unique<VBoxManageRunner>(arguments);
+    QObject::connect(runner.get(), &VBoxManageRunner::done, context, functor);
+    commandQueue()->enqueue(std::move(runner));
+}
+
+void VBoxVirtualMachinePrivate::doSetReservedPortForwarding(ReservedPort which, quint16 port,
+        const QObject *context, const Functor<bool> &functor)
+{
+    Q_Q(VBoxVirtualMachine);
+    Q_ASSERT(context);
+    Q_ASSERT(functor);
+
+    qCDebug(vms) << "Setting reserved port forwarding" << which << "for" << q->uri().toString()
+        << "to" << port;
+
+    QString ruleName;
+    int guestPort{};
```

Comment:
Technically, this works, but how is it better than "int guestPort = 0;", which would do the same thing and would also be clear to the reader?

Reply:
it's the same as "QString ruleName;", but for primitive types you need to require the default construction explicitly and at the same time "int questPort();" would be treated as forward declaration of a function questPort. That's one of the reasons why the unified initialization was introduced with C++11. If I do not need a specific value, I prefer expressing it using default initialization, explicit where it is needed.

Reply:
While I do understand your logic, I have a minor issue with the fact that you are using the uniform initialization to perform the default initialization to zero. Yes, I do understand that you are trying to express the fact that you don't really care about the value, but at the same time you are hiding the fact that it is being initialized to zero. Sure, not a big issue for anyone who can remember all of the C++ standard by heart.